### ISSUE TITLE
Logged connection endpoint contains wrong IP address

### ISSUE THREAD

**Issue by CarstenGrohmann** [neutrinolabs/xrdp#381]
The some of the xrdp log messages after closing a connection wrong resp. non-existing IP address. The example below shows the IP address 109.112.47.46 that doesn't exists in my network.

I checked some of the xrdp-related processes and don't find any TCP connection with file descriptor 12 or 13. But both file descriptors are in use for Unix domain sockets. 

Maybe the garbage come from using `getpeername()` on a Unix domain socket.

Please adapt the code to show the right peer name / end in such cases.

Extract from `/var/log/xrdp.log`:
```
# tail /var/log/xrdp.log
[20160531-14:32:12] [INFO ] The following channel is allowed: drdynvc (5)
[20160531-14:32:12] [DEBUG] The allow channel list now initialized for this session
[20160531-14:34:24] [INFO ] An established connection closed to endpoint: 109.112.47.46:12148 - socket: 13
[20160531-14:34:24] [INFO ] An established connection closed to endpoint: 192.168.1.21:59727 - socket: 8
[20160531-14:34:24] [DEBUG] xrdp_mm_module_cleanup
[20160531-14:34:24] [INFO ] An established connection closed to endpoint: 109.112.47.46:12148 - socket: 12
[20160531-14:34:24] [ERROR] Listening socket is in wrong state we terminate listener
```

Open files from xrdp-related processes filtered by fd 12 and fd 13:
```
xrdp      22931     root   12u  unix 0xffff881025c10b40       0t0   255532 socket
xrdp      22931     root   13u  unix 0xffff881025c25640       0t0   249600 socket
xrdp-chan 24939 carsten 12u  unix 0xffff881025c27800       0t0   257545 /tmp/.xrdp/xrdp_chansrv_socket_10
xrdp-chan 24939 carsten 13u  unix 0xffff8810244221c0       0t0   258526 socket
```

Open TCP connections:
```
# grep IPv4 trace
xrdp-sesm  1206     root    7u  IPv4               8957       0t0      TCP localhost:findviatv (LISTEN)
xrdp      10052     root    7u  IPv4              85695       0t0      TCP *:ms-wbt-server (LISTEN)
xrdp      22931     root    8u  IPv4             217345       0t0      TCP mysrv:ms-wbt-server->myclient:59727 (ESTABLISHED)
```

Thank you,
Carsten

**Comment by CarstenGrohmann** [comment]
The message occured after I opened and closed a Gnome session via the Session Manager.

**Comment by metalefty** [comment]
This is a long life known bug. Haven't fixed yet.

http://blog.gmane.org/gmane.network.rdesktop.xrdp.devel/page=20
https://sourceforge.net/p/xrdp/discussion/389417/thread/e8fb6b34/

**Comment by CarstenGrohmann** [comment]
I've a suggestion to review:
```C++
void APP_CC
g_write_ip_address(int rcv_sck, char *ip_address, int bytes)
{
    struct sockaddr_in sin;
    struct sockaddr_un sun;
    struct in_addr in;
#if defined(_WIN32)
    int len;
#else
    unsigned int len;
#endif

    len = sizeof(sun);
    memset(&sun, 0, len);

    /* Get address family and local address the socket is bound to */
    if (getsockname(rcv_sck, (struct sockaddr *)&sun, &len) < 0) {
        /* Common error: EBADF */
        log_message(LOG_LEVEL_DEBUG,
                    "getsockname() for fd %d failed: %d (%s)",
                    rcv_sck, g_get_errno(), g_get_strerror());
        goto error;
	}

    switch (sun.sun_family) {
#ifndef _WIN32
        case AF_UNIX:
            /* All OS: unnamed socket */
            if (len == sizeof(sa_family_t))
            {
                snprintf(ip_address, bytes,
                         "Unnamed socket - fd: %d", rcv_sck);
            }
            /* Abstract socket - Linux specific and probably Solaris */
            else if ( len > sizeof(sa_family_t) && sun.sun_path[0] == '\0')
            {
                 snprintf(ip_address, bytes,
                         "Abstract Linux socket - fd: %d", rcv_sck);
            }
            /* All OS: path name */
            else
            {
                snprintf(ip_address, bytes,
                         "Local socket file %s - fd: %d", sun.sun_path, rcv_sck);
            }
            return;
#endif
        case AF_INET:
        case AF_INET6:
            len = sizeof(sin);
            memset(&sin, 0, len);
            /* Get peer connection details */
            if (getpeername(rcv_sck, (struct sockaddr *)&sin, &len) < 0)
            {
                /* Common error: ENOTCONN */
                log_message(LOG_LEVEL_DEBUG,
                            "getpeername() for fd %d failed: %d (%s)",
                            rcv_sck, g_get_errno(), g_get_strerror());
                goto error;
            }

            memset(&in, 0, sizeof(in));
            in.s_addr = sin.sin_addr.s_addr;
            snprintf(ip_address, bytes, "%s:%d - fd: %d",
                     inet_ntoa(in), ntohs(sin.sin_port), rcv_sck);
            return;

        default:
            log_message(LOG_LEVEL_DEBUG,
                        "Unknown address family %d for fd: %d",
                        sun.sun_family, rcv_sck);
    }

error:
    snprintf(ip_address, bytes, "NULL:NULL - fd: %d", rcv_sck);
}
```

Maybe the function could be renamed to something like `g_write_socket_details` and `NULL:NULL` could be changed to `unkown socket details`.

**Comment by metalefty** [comment]
Tested but I can see no client ip address (192.168.69.1).

```
[20160628-16:22:59] [INFO ] An established connection closed to endpoint: Abstract Linux socket - fd: 19
[20160628-16:23:03] [INFO ] An established connection closed to endpoint: Abstract Linux socket - fd: 19
[20160628-16:23:07] [INFO ] An established connection closed to endpoint: Abstract Linux socket - fd: 19
[20160628-16:23:11] [INFO ] An established connection closed to endpoint: Abstract Linux socket - fd: 19
[20160628-16:23:14] [INFO ] A connection received from: 0.0.0.0 port 56913
[20160628-16:23:14] [INFO ] An established connection closed to endpoint: 0.0.0.0:56913 - fd: 11
[20160628-16:23:14] [DEBUG] getpeername() for fd 10 failed: 57 (Socket is not connected)
[20160628-16:23:14] [INFO ] An established connection closed to endpoint: NULL:NULL - fd: 10
[20160628-16:23:14] [DEBUG] getpeername() for fd 11 failed: 57 (Socket is not connected)
[20160628-16:23:14] [INFO ] An established connection closed to endpoint: NULL:NULL - fd: 11
xrdp_mcs_disconnect - socket closed
[20160628-16:23:14] [ERROR] Listening socket is in wrong state we terminate listener
[20160628-16:23:14] [INFO ] A connection received from: 0.0.0.0 port 56914
[20160628-16:23:14] [INFO ] An established connection closed to endpoint: 0.0.0.0:56914 - fd: 11
[20160628-16:23:14] [DEBUG] getpeername() for fd 10 failed: 57 (Socket is not connected)
[20160628-16:23:14] [INFO ] An established connection closed to endpoint: NULL:NULL - fd: 10

```

**Comment by metalefty** [comment]
Fixed by #411.

**Comment by CarstenGrohmann** [comment]
@proski @metalefty: "Thank you" for fixing this bug!
